File: Python/FreiStat/Data_storage/data_software_storage.py
"""
Module implementing a class for handling the data which is required for the 
FreiStat software to operate.

This class holds reference to all other main objects, which are required by the 
FreiStat Python, in order to operate.

"""

__author__ = "Mark Jasper"
__contact__ = "University of Freiburg, IMTEK, Jochen Kieninger"
__credits__ = "Mark Jasper"

__version__ = "1.0.0"
__maintainer__ = "Mark Jasper"
__email__ = "mark.jasper@imtek.uni-freiburg.de, kieninger@imtek.uni-freiburg.de"

# Import dependencies

# Import internal dependencies
from .constants import *


# Forward declaration
def Communication():
    pass


def JSON_Parser():
    pass


def JSON_Telegram_Generator():
    pass


def DataHandling():
    pass


class DataSoftwareStorage:
    """
    Description
    -----------
    Class which stores all references and general variables for the Python
    library of the FreiStat software.

    """

    def __init__(self) -> None:
        """
        Description
        -----------
        Constructor of class DataSoftwareStorage

        """
        # Initalize class variables
        self._dataHandling = None
        self._jsonParser = None
        self._jsonTelegramGenerator = None

        self._bLowPerformanceMode: bool = False

        self._systemStatus: int = FREISTAT_BOOTUP

    # Setter methods
    def set_SystemStatus(self, iSystemStatus: int) -> None:
        """
        Description
        -----------
        Save system status. The following status modes are defined*:

        FREISTAT_BOOTUP         = 0             # FreiStat is in boot up process
        FREISTAT_WAITING        = 1             # FreiStat in waiting state
        FREISTAT_EXP_STARTED    = 2             # Start experiment
        FREISTAT_EXP_RUNNING    = 3             # Experiment running
        FREISTAT_EXP_COMPLETED  = 4             # Experiment completed
        FREISTAT_EXP_CANCELED   = 5             # Experiment canceled

        *Taken from constants.py

        Parameters
        ----------
        `iSystemStatus` : int
            Status of the system encoded in an integer value.

        """
        self._systemStatus = iSystemStatus

    def set_LowPerformanceMode(self, bLowPerformanceMode: bool) -> None:
        """
        Description
        -----------
        Save low performanance mode flag

        Parameters
        ----------
        `bLowPerformanceMode ` bool
            Flag indicating if the low performance mode is enabled or not.

        """
        self._bLowPerformanceMode = bLowPerformanceMode

    def setJSON_Parser(self, jsonParser: JSON_Parser) -> None:
        """
        Description
        -----------
        Set reference of the JSON parser object.

        Parameters
        ----------
        `jsonParser` : JSON_Parser
            Reference to the JSON parser object

        """
        self._jsonParser = jsonParser

    def setCommunication(self, communication: Communication) -> None:
        """
        Description
        -----------
        Set reference of the serial communication object.

        Parameters
        ----------
        `communication` : Communication
            Reference to the serial communication object

        """
        self._communication = communication

    def setDataHandling(self, dataHandling: DataHandling) -> None:
        """
        Description
        -----------
        Set reference of the data handling object.

        Parameters
        ----------
        `dataHandling` : DataHandling
            Reference to the data handling object

        """
        self._dataHandling = dataHandling

    def setJSON_TelegramGenerator(
        self, jsonTelegramGenerator: JSON_Telegram_Generator
    ) -> None:
        """
        Description
        -----------
        Set reference of the JSON telegram generator object.

        Parameters
        ----------
        `jsonTelegramGenerator` : JSON_Telegram_Generator
            Reference to the JSON telegram generator object

        """
        self._jsonTelegramGenerator = jsonTelegramGenerator

    # Getter methods
    def get_SystemStatus(self) -> int:
        """
        Description
        -----------
        Get system status. See setter method or constant.py for possible values.

        Return
        ------
        `iSystemStatus` : int
            Status of the system encoded in an integer value

        """
        return self._systemStatus

    def get_LowPerformanceMode(self) -> bool:
        """
        Description
        -----------
        Get low performanance mode flag

        Return
        ------
        `bLowPerformanceMode ` bool
            Flag indicating if the low performance mode is enabled or not.

        """
        return self._bLowPerformanceMode

    def getJSON_Parser(self) -> JSON_Parser:
        """
        Description
        -----------
        Get reference of the JSON parser object.

        Return
        ------
        `jsonParser` : JSON_Parser
            Reference to the JSON parser object

        """
        return self._jsonParser

    def getCommunication(self) -> Communication:
        """
        Description
        -----------
        Get reference of the serial Communication object.

        Return
        ------
        `communication` : Communication
            Reference to the serial Communication object

        """
        return self._communication

    def getDataHandling(self) -> DataHandling:
        """
        Description
        -----------
        Get reference of the data handling object.

        Return
        ----------
        `dataHandling` : DataHandling
            Reference to the data handling object

        """
        return self._dataHandling

    def getJSON_TelegramGenerator(self) -> JSON_Telegram_Generator:
        """
        Description
        -----------
        Get reference of the JSON telegram generator object.

        Return
        ------
        `jsonTelegramGenerator`: JSON_Telegram_Generator
            Reference to the JSON telegram generator object

        """
        return self._jsonTelegramGenerator

File: src/data_storage/data_software_storage.cpp
/******************************************************************************
 * @brief: Source file for the data storage which stores all required data for
 * the FreiStat and holds references to all other objects.
 * 
 * @author: Mark Jasper
 * @version: V 1.5.0
 * @date: 13.09.2021
 * 
 *****************************************************************************/

// Include guard
#ifndef data_software_storage_CPP
#define data_software_storage_CPP

// Include headers
#include "data_software_storage.h"

/******************************************************************************
 * @brief Default constructor of the class C_DataSoftwareStorage
 * 
 *****************************************************************************/ 
C_DataSoftwareStorage::C_DataSoftwareStorage(){}

/******************************************************************************
 * @brief Starting method for the class C_DataSoftwareStorage
 * 
 *****************************************************************************/
void C_DataSoftwareStorage::Begin(C_AD5940_Setup * c_AD5940_Setup){
    // Save reference to AD5940 setup object
    c_AD5940_Setup_ = c_AD5940_Setup;
    
    // Initialize system status
    iSystemStatus_ = FREISTAT_BOOTUP;

    // Initialize sequence status
    iSequenceStatus_ = FREISTAT_EXS_DISABLED;
}

/******************************************************************************
 * Setter methods
 *****************************************************************************/
/******************************************************************************
 * @brief Setter method for saving reference pointer to the general data storage 
 * object
 * @param c_DataStorageGeneral: pointer to the genral data storage object
 *****************************************************************************/
void C_DataSoftwareStorage::set_DataStorageGeneral(C_DataStorageGeneral * 
        c_DataStorageGeneral){
    c_DataStorageGeneral_ = c_DataStorageGeneral;
}

/******************************************************************************
 * @brief Setter method for saving reference pointer to the local data storage 
 * object.
 * @param c_DataStorageLocal: pointer to the local data storage object
 *****************************************************************************/
void C_DataSoftwareStorage::set_DataStorageLocal(C_DataStorageLocal * 
        c_DataStorageLocal){
    c_DataStorageLocal_ = c_DataStorageLocal;
}

/******************************************************************************
 * @brief Setter method for saving reference pointer to JSON parser object
 * @param c_JSONParser: pointer to JSON parser object
 *****************************************************************************/
void C_DataSoftwareStorage::set_JSONParser(C_JSONParser * c_JSONParser){
    c_JSONParser_ = c_JSONParser;
}

/******************************************************************************
 * @brief Setter method for saving reference pointer to communication object
 * @param c_Communication: pointer to Communication object
 *****************************************************************************/
void C_DataSoftwareStorage::set_Communication(C_Communication * c_Communication){
    c_Communication_ = c_Communication;
}

/******************************************************************************
 * @brief Setter method for saving reference pointer to electrochemical method
 * object
 * @param c_ElectrochemicalMethod: pointer to electrochemical method object
 *****************************************************************************/
void C_DataSoftwareStorage::set_ElectrochemicalMethod(C_ElectrochemicalMethod * 
    c_ElectrochemicalMethod){
    c_ElectrochemicalMethod_ = c_ElectrochemicalMethod;
}

/******************************************************************************
 * @brief Method for changing the saved system status
 * @param iSystemStatus: encoded system status
 *****************************************************************************/
void C_DataSoftwareStorage::set_SystemStatus(int iSystemStatus){
    iSystemStatus_ = iSystemStatus;
}

/******************************************************************************
 * @brief Method for changing the saved sequence status
 * @param iSequenceStatus: encoded sequence status
 *****************************************************************************/
void C_DataSoftwareStorage::set_SequenceStatus(int iSequenceStatus){
    iSequenceStatus_ = iSequenceStatus;
}

/******************************************************************************
 * @brief Method for changing the saved control status
 * @param iControlStatus: encoded control status
 *****************************************************************************/
void C_DataSoftwareStorage::set_ControlStatus(int iControlStatus){
    iControlStatus_ = iControlStatus;
}

/******************************************************************************
 * Getter methods
 *****************************************************************************/
/******************************************************************************
 * @brief Getter method for returning reference pointer to AD5940 setup object
 * @return: Returns pointer to AD5940 setup object
 *****************************************************************************/
C_AD5940_Setup * C_DataSoftwareStorage::get_AD5940Setup(){
    return c_AD5940_Setup_;
}

/******************************************************************************
 * @brief Getter method for returning reference pointer to the general data 
 * storage object
 * @return: Returns pointer to the general data storage object
 *****************************************************************************/
C_DataStorageGeneral * C_DataSoftwareStorage::get_DataStorageGeneral(){
    return c_DataStorageGeneral_;
}

/******************************************************************************
 * @brief Getter method for returning reference pointer to the local data 
 * storage object
 * @return: Returns pointer to the local data storage object
 *****************************************************************************/
C_DataStorageLocal * C_DataSoftwareStorage::get_DataStorageLocal(){
    return c_DataStorageLocal_;
}

/******************************************************************************
 * @brief Getter method for returning reference pointer to JSON parser object
 * @return: Returns pointer to JSON parser object
 *****************************************************************************/
C_JSONParser * C_DataSoftwareStorage::get_JSONParser(){
    return c_JSONParser_;
}

/******************************************************************************
 * @brief Getter method for returning reference pointer to communication object
 * @return: Returns pointer to communication object
 *****************************************************************************/
C_Communication * C_DataSoftwareStorage::get_Communication(){
    return c_Communication_;
}

/******************************************************************************
 * @brief Getter method for returning reference pointer to electrochemical 
 * method object
 * @return: Returns pointer to electrochemical method object
 *****************************************************************************/
C_ElectrochemicalMethod * C_DataSoftwareStorage::get_ElectrochemicalMethod(){
    return c_ElectrochemicalMethod_;
}

/******************************************************************************
 * @brief Method for returning current system status
 * @details The system status is encoded in an integer value, from which the 
 * following are defined:
 * System status : Meaning
 * 0 : FreiStat in boot up process
 * 1 : FreiStat in waiting state
 * 2 : Experiment type received, waiting for experiment parameters
 * 3 : Experiment parameters received, completing setup now
 * 4 : FreiStat ready for running experiment, waiting for start telegram
 * 5 : Experiment running
 * 6 : Experiment completed
 * 7 : Experiment canceled
 * @return: Returns encoded system status
 *****************************************************************************/
int C_DataSoftwareStorage::get_SystemStatus(){
    return iSystemStatus_;
}

/******************************************************************************
 * @brief Method for returning the saved sequence status
 * @details The sequenece status is encoded in an integer value, from which the 
 * following are defined:
 * Sequence status  :   Meaning
 * 0                :   Not Defiend
 * 1                :   Enabled
 * 2                :   Disabled
 * @return: encoded sequence status
 *****************************************************************************/
int C_DataSoftwareStorage::get_SequenceStatus(){
    return iSequenceStatus_;
}

/******************************************************************************
 * @brief Method for returning current control status
 * @details The control status is encoded in an integer value, from which the 
 * following are defined:
 * Control status   :   Meaning 
 * 0                :   Waiting
 * 1                :   Start
 * 2                :   Stopp
 * @return: Returns encoded system status
 *****************************************************************************/
int C_DataSoftwareStorage::get_ControlStatus(){
    return iSystemStatus_;
}

#endif  /* data_software_storage_CPP */